src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -404,9 +404,44 @@ internal override bool ApplyNullableTransforms(byte defaultTransformFlag, Immuta
             return true;
         }
 
-        internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnotations, TypeWithAnnotations> transform)
+        internal override TypeSymbol SetObliviousNullabilityForReferenceTypes()
         {
-            return WithElementType(transform(ElementTypeWithAnnotations));
+            if (ElementType.TypeKind == TypeKind.Array)
+            {
+                // The language supports deeply nested arrays, up to 10,000+ instances. This means we can't implemented a head
+                // recursive solution here. Need to take an iterative approach to building up the annotations here.
+                var builder = ArrayBuilder<ArrayTypeSymbol>.GetInstance();
+
+                // Dig through the element types to get the set of ArrayTypeSymbol instances that need to be marked
+                // as oblivious
+                var array = this;
+                TypeWithAnnotations elementType;
+                do
+                {
+                    builder.Add(array);
+                    elementType = array.ElementTypeWithAnnotations;
+                    array = elementType.TypeKind == TypeKind.Array
+                        ? (ArrayTypeSymbol)elementType.Type
+                        : null;
+                }
+                while (!(array is null));
+
+                // Fixup the element type on the most nested array
+                elementType = elementType.SetObliviousNullabilityForReferenceTypes();
+                array = (builder[builder.Count - 1]).WithElementType(elementType);
+
+                // All but the most nested array instance is just an oblivious array with the same custom modifiers
+                for (var i = builder.Count - 2; i >= 0; i--)
+                {
+                    var previousArray = builder[i];
+                    elementType = TypeWithAnnotations.Create(array, NullableAnnotation.Oblivious, previousArray.ElementTypeWithAnnotations.CustomModifiers);
+                    array = (ArrayTypeSymbol)(previousArray.WithElementType(elementType));
+                }
+
+                builder.Free();
+                return array;
+            }
+            return WithElementType(ElementTypeWithAnnotations.SetObliviousNullabilityForReferenceTypes());
```

Comment:
>WithElementType [](start = 19, length = 15)

Is this different than the base case in the iterative approach above? In short, do we need to differentiate the case where the top-level element type is not an array?